### ISSUE TITLE
Use EasyConfigs of installed modules

### ISSUE THREAD

**Issue by Flamefire** [easybuilders/easybuild-framework#3128]
We recently had the following situation:

- CUDA included in fosscuda/2019a toolchain did not work on our system (at all)
- Write updated CUDA EC using newer CUDA
- Upgraded fosscuda/2019a EC to use newer CUDA EC
- Install Software X with this fosscuda/2019a
- Some other user tried to install Software Y (also with fosscuda/2019a) but installation failed due to the updated CUDA EC was not found in the repo

So basically:
- There is a matching module
- it contains the used EC
- EB is able to find the module
- EB knows where the used EC is under that module (because EB was used to put it there)

So why doesn't EB use it then? IMO those should even be preferred to avoid an inconsistently configured system:
- EB version x uses ECs with dependency Y
- EB version x+1 updates ECs to use dependency Y+1 (e.g. due to bugfixes as happened to TensorFlow)

--> Software installed before and after the EB upgrade have different requirements.

**Comment by ocaisa** [comment]
This is configurable behaviour, see [here](https://easybuild.readthedocs.io/en/latest/Using_the_EasyBuild_command_line.html#searching-for-easyconfigs-the-robot-search-path).

> So why doesn't EB use it then? IMO those should even be preferred to avoid an inconsistently configured system

You can configure that by setting the `robot` option to to point to your `eb_repo` path, it is just not the default. This is a site-specific wish, on our site we want to maintain complete control over the dependencies, so in the production stack we use a "golden" repository to resolve dependencies. In our "Devel" stack we allow people (in our `software` group) to resolve using the installed modules in that stack. We also introduced the option to allow them to search in the EasyBuild shipped easyconfigs but not resolve dependencies from there ( `--search-paths`, documented in the help commmand but doesn't seem to be in the docs). In other words, all of this is configurable behaviour.

**Comment by Flamefire** [comment]
> You can configure that by setting the robot option to to point to your eb_repo path

What is the `eb_repo` path? I was referring to the `<module_install_path>/<module>/easybuild/<module>.eb` which is installed alongside every EB installed module. There is no single path to be added and I haven't found anything in the linked docs. To quote:

> For each dependency that does not have a matching module installed yet, EasyBuild requires a corresponding easyconfig file. [...]
> Note: Sometimes easyconfig files are needed even when the modules for the dependencies are already available [...]

Nothing mentioning the ECs of installed modules.

>  in the production stack we use a "golden" repository to resolve dependencies. In our "Devel" stack we allow people (in our software group) to resolve using the installed modules in that stack. 

I don't fully understand this. What I'm proposing is that for the case an EC is `needed even when the modules for the dependencies are already available` then that EC should first come from that module.

I don't see that conflicting with your setup. In the production stack you ignore installed modules (or please clarify) so this does not apply and in the dev stack resolved modules have their "blessed" ECs with them.

**Comment by ocaisa** [comment]
Take a look at https://easybuild.readthedocs.io/en/latest/Configuration.html#easyconfigs-repo . For every piece of software installed you get a new entry there, if you add that as an argument to `--robot` then you will get your desired behaviour (things are resolved there first).

**Comment by Flamefire** [comment]
Ok, thanks. This doesn't change the question: Why not use the ECs of the modules?

While the repo-path is good it has drawbacks:
- default in users home means the ECs on our cluster are not in 1 location
- failed copies are non-fatal hence those ECs might be incomplete
- duplicated ECs as those ECs are already in the modules
- might become out-of-sync with installed modules

**Comment by ocaisa** [comment]
@boegel I think it is worth raising/debating this at our next confcall. Since 4.x this is now more feasible that it was in 3.x since we now have an installation default that is guaranteed to be unique (which, previously, was not true for HMNS).

> * default in users home means the ECs on our cluster are not in 1 location

Sure, but you are suggesting then that they are using non-defaults for other parameters to install on the cluster (otherwise installations would also be in their home folders)...why not just add this to the local (user) configuration?

> * failed copies are non-fatal hence those ECs might be incomplete

Hard to argue with this (we handle this through group permissions).

> * duplicated ECs as those ECs are already in the modules

Right, but this is a design choice to make it trivial for people to add that repository to their search path. What is the expectation when eb does a search for easyconfigs (for example), that it will search the entire installation tree for potential matches?

> * might become out-of-sync with installed modules

I think that's already implied by the second point.

**Comment by Flamefire** [comment]
> Sure, but you are suggesting then that they are using non-defaults for other parameters to install on the cluster (otherwise installations would also be in their home folders)...why not just add this to the local (user) configuration?

My point was: Multiple users (admins) take care of installing modules onto the cluster out of (e.g.) the official sources. If one admin has to change or add an EC to install a module, that EC will be added to **his** home directory. If another user/admin wants to install a module based on that module, EB will not find the EC although it (can) know where it is: Right next to the module.

>> might become out-of-sync with installed modules
>
> I think that's already implied by the second point.

See above: It can also happen if the repository path is left as a default for one of the admins. E.g. forgot to change, after an update, ...

IMO the installed modules ECs are the only, reliable source of truth. This are the ones that were used to install the module. An EC from the repository path may or may not have been used.

**Comment by akesandgren** [comment]
"If one admin has to change or add an EC to install a module, that EC will be added to his home directory."

Then you're using a bad methodology.
The site should have a central repository for changed and new (not yet upstream) easyconfigs, easyblocks and easyhooks.
Using a git repo is preferable here.
Then you point that one out with
robot-paths=/hpc2n/eb/custom/easyconfigs:%(DEFAULT_ROBOT_PATHS)s
in the central easyconfig config file, including the default paths from the easybuild installation.

**Comment by ocaisa** [comment]
@Flamefire I agree that this is worth exploring but an implementation is not something I would do without a lot of thought, changing the priority in how dependency resolution is done to resolve to an ec that might not even be searchable is, I would say, not a good idea (of course this is a solvable problem, but I would be wary that there are more issues to consider).

The issue you ran into is actually more to do with toolchains rather than dependencies, eb doesn't care about the easyconfig content if it can find the module, but the toolchain definition it does need. In that case, I think it is hard to argue that what is defined on the system is not what should be respected.
